temporal-server/authorizer/claimer.go

```diff
@@ -0,0 +1,245 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package authorizer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/canonical/ofga"
+	"go.temporal.io/server/common/authorization"
+	"go.uber.org/zap"
+)
+
+//go:generate mockgen -destination=mocks/groups_provider_gen.go -package=mock github.com/canonical/charmed-temporal-image/temporal-server/authorizer NamespaceAccessProvider,TokenVerifier
+
+type NamespaceAccess struct {
+	Namespace string
+	Relation  string
+}
+
+// NamespaceAccessProvider is an interface that defines the method to retrieve namespace
+// access information for a given email.
+type NamespaceAccessProvider interface {
+	GetNamespaceAccessInformation(ctx context.Context, email string) ([]NamespaceAccess, error)
+}
+
+// AuthClient implements the necessary methods needed to fetch namespace access
+// information from an OpenFGA store.
+type AuthClient struct {
+	OfgaClient *ofga.Client
+}
+
+// TokenClaimMapper implements Temporal authorization.ClaimMapper,
+// performing a OIDC-based authorization.
+type TokenClaimMapper struct {
+	// TokenVerifier is used to verify the provided access token.
+	TokenVerifier TokenVerifier
+	// NamespaceAccessProvider is used to identify the namespaces that the user
+	// logging in via the access token has access to.
+	NamespaceAccessProvider NamespaceAccessProvider
+	// AdminGroups is a comma-separated list of groups which gives full system
+	// access to all users belonging to them.
+	AdminGroups string
+	// OpenAccessNamespaces is a comma-separated list of namespaces to which everyone
+	// with valid login credentials has access. If empty, no such namespace will be
+	// configured.
+	OpenAccessNamespaces string
+	// Logger is used for logging TokenClaimMapper operations.
+	Logger *zap.Logger
```

Comment:
IIRC, we usually rely on zap's global logger in our services, it's not necessary to pass it usually (or at least I've never seen it passed around in our projects).